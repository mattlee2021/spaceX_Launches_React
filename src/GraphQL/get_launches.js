import { gql } from '@apollo/client'

/**
 * Query used to fetch data from SpaceX GraphQL API.
 * This Query is used in LaunchesContainer.
 */
export const GET_LAUNCHES = gql`
query launchesPast($limit: Int, $filter_Mission:String, $filter_launchSite:String, $sort_by:String) {
    launchesPast(limit:$limit, find:{mission_name:$filter_Mission, site_name_long:$filter_launchSite}, sort:$sort_by){
        id
        mission_name
        launch_date_local
        launch_site {
            site_name_long
        }
        links {
            video_link
        }
        rocket {
            rocket_name
        }
    }
}`
