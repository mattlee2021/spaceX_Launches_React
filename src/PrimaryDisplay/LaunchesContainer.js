import React from "react";
import { useQuery } from "@apollo/client";
import { GET_LAUNCHES } from "../GraphQL/get_launches";
import Launch from "./Launch";
import styles from "./LaunchesContainer.module.css";

const LaunchesContainer = (props) => {
  const { data, loading, error } = useQuery(GET_LAUNCHES, {
    variables: {
      limit: 10,
      filter_Mission: props.search,
      filter_launchSite: props.launchSite_filter,
      sort_by: props.sort,
    },
  });

  if (loading) {
    return <div className={styles.processing__text}>Loading...</div>;
  } else if (error) {
    return (
      <div className={styles.processing__text}>
        Error... Please Refresh The Page
      </div>
    );
  } else {
    return (
      <div className={styles.launch__container}>
        {data.launchesPast.length > 0 ? (
          data.launchesPast.map((launchData) => (
            <Launch key={launchData.id} launch={launchData} />
          ))
        ) : (
          <div className={styles.processing__text}> No Launches Found </div>
        )}
      </div>
    );
  }
};

export default LaunchesContainer;
