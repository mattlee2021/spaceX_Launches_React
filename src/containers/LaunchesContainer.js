import React from 'react';
import { useQuery } from '@apollo/client';
import { GET_LAUNCHES } from '../graphql/get_launches';
import { Launches } from '../components/DisplayLaunches';

/**
 * Fetching 10 SpaceX Launches with limit:10 that is passed to GET_LAUNCHES.
 * Displaying either Loading and Error messages that are passed from useQuery()
 * If there are no errors and data has loaded, I am mapping each subset of launch data 
 * to the <Launches /> component. 
 */
export function LaunchesContainer(props) {
    console.log("Searched: ", props.search, "; Filter is ", props.launchSite_filter, "; Sort By ", props.sort)
    const { data, loading, error } = useQuery(GET_LAUNCHES, {
        variables: {
            limit: 20, filter_Mission: props.search,
            filter_launchSite: props.launchSite_filter, sort_by: props.sort
        },
    });

    if (loading) {
        return (<div className="Processing">Loading...</div>)
    }

    if (error) {
        console.error(error)
        return (<div className="Processing">Error. Check the console.</div>)
    }

    return (
        <div className="container">
            {data.launchesPast && data.launchesPast.map(launch_data =>
                <Launches key={launch_data.id} launch={launch_data} />)}

        </div>
    )

}