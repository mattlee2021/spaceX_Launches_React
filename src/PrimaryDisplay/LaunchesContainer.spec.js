import TestRenderer from "react-test-renderer";
import LaunchesContainer from "./LaunchesContainer";
import { MockedProvider } from "@apollo/client/testing";
import { GET_LAUNCHES } from "../GraphQL/get_launches";
import { render, screen } from "@testing-library/react";
import { GraphQLError } from "graphql";

describe("LaunchesContainer", () => {
  const mockLaunchesContainerData = {
    launchesPast: [
      {
        id: "109",
        mission_name: "Starlink-15 (v1.0)",
        launch_date_utc: "2020-10-24T15:31:00.000Z",
        launch_site: {
          site_name_long:
            "Cape Canaveral Air Force Station Space Launch Complex 40",
        },
        links: {
          video_link: "https://youtu.be/J442-ti-Dhg",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "108",
        mission_name: "Sentinel-6 Michael Freilich",
        launch_date_utc: "2020-11-21T17:17:00.000Z",
        launch_site: {
          site_name_long: "Vandenberg Air Force Base Space Launch Complex 4E",
        },
        links: {
          video_link: "https://youtu.be/aVFPzTDCihQ",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "107",
        mission_name: "Crew-1",
        launch_date_utc: "2020-11-16T00:27:00.000Z",
        launch_site: {
          site_name_long: "Kennedy Space Center Historic Launch Complex 39A",
        },
        links: {
          video_link: "https://youtu.be/bnChQbxLkkI",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "106",
        mission_name: "GPS III SV04 (Sacagawea)",
        launch_date_utc: "2020-11-05T23:24:00.000Z",
        launch_site: {
          site_name_long:
            "Cape Canaveral Air Force Station Space Launch Complex 40",
        },
        links: {
          video_link: "https://youtu.be/wufXF5YKR1M",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "105",
        mission_name: "Starlink-14 (v1.0)",
        launch_date_utc: "2020-10-24T15:31:00.000Z",
        launch_site: {
          site_name_long:
            "Cape Canaveral Air Force Station Space Launch Complex 40",
        },
        links: {
          video_link: "https://youtu.be/2gbVgTxLgN0",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "104",
        mission_name: "Starlink-13 (v1.0)",
        launch_date_utc: "2020-10-18T12:25:00.000Z",
        launch_site: {
          site_name_long: "Kennedy Space Center Historic Launch Complex 39A",
        },
        links: {
          video_link: "https://youtu.be/UM8CDDAmp98",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "103",
        mission_name: "Starlink-12 (v1.0)",
        launch_date_utc: "2020-10-06T11:29:00.000Z",
        launch_site: {
          site_name_long: "Kennedy Space Center Historic Launch Complex 39A",
        },
        links: {
          video_link: "https://youtu.be/8O8Z2yPyTnc",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "102",
        mission_name: "Starlink-11 (v1.0)",
        launch_date_utc: "2020-09-03T12:46:00.000Z",
        launch_site: {
          site_name_long: "Kennedy Space Center Historic Launch Complex 39A",
        },
        links: {
          video_link: "https://youtu.be/_j4xR7LMCGY",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "101",
        mission_name: "SAOCOM 1B, GNOMES-1, Tyvak-0172",
        launch_date_utc: "2020-08-30T23:18:00.000Z",
        launch_site: {
          site_name_long:
            "Cape Canaveral Air Force Station Space Launch Complex 40",
        },
        links: {
          video_link: "https://youtu.be/P-gLOsDjE3E",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "100",
        mission_name: "Starlink-10 (v1.0) & SkySat 19-21",
        launch_date_utc: "2020-08-18T14:31:00.000Z",
        launch_site: {
          site_name_long:
            "Cape Canaveral Air Force Station Space Launch Complex 40",
        },
        links: {
          video_link: "https://youtu.be/jTMJK7wb0rM",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
      {
        id: "99",
        mission_name: "Starlink-9 (v1.0) & BlackSky Global 5-6",
        launch_date_utc: "2020-08-07T05:12:00.000Z",
        launch_site: {
          site_name_long: "Kennedy Space Center Historic Launch Complex 39A",
        },
        links: {
          video_link: "https://youtu.be/KU6KogxG5BE",
        },
        rocket: {
          rocket_name: "Falcon 9",
        },
      },
    ],
  };

  const mockData = {
    request: {
      query: GET_LAUNCHES,
      variables: {
        limit: 10,
        filter_Mission: "",
        filter_launchSite: "",
        sort_by: "",
      },
    },
    result: {
      data: mockLaunchesContainerData,
    },
  };

  it("Should display Loading...", () => {
    render(
      <MockedProvider mocks={[mockData]} addTypename={false}>
        <LaunchesContainer search={""} launchSite_filter={""} sort={""} />
      </MockedProvider>
    );
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("Should display the mocked data", async () => {
    const testRenderer = TestRenderer.create(
      <MockedProvider mocks={[mockData]} addTypename={false}>
        <LaunchesContainer search={""} launchSite_filter={""} sort={""} />
      </MockedProvider>
    );
    testRenderer.toJSON();
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(testRenderer).toMatchSnapshot();
  });

  it("Should display the error message on network error", async () => {
    const mockError = {
      request: {
        query: GET_LAUNCHES,
        variables: {
          limit: 10,
          filter_Mission: "",
          filter_launchSite: "",
          sort_by: "",
        },
      },
      error: new Error("An error occurred"),
    };

    render(
      <MockedProvider mocks={[mockError]} addTypename={false}>
        <LaunchesContainer search={""} launchSite_filter={""} sort={""} />
      </MockedProvider>
    );
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(
      screen.getByText("Error... Please Refresh The Page")
    ).toBeInTheDocument();
  });

  it("Should display the error message on GraphQL error", async () => {
    const mockError = {
      request: {
        query: GET_LAUNCHES,
        variables: {
          limit: 10,
          filter_Mission: "",
          filter_launchSite: "",
          sort_by: "",
        },
      },
      result: { errors: [new GraphQLError("Error")] },
    };

    render(
      <MockedProvider mocks={[mockError]} addTypename={false}>
        <LaunchesContainer search={""} launchSite_filter={""} sort={""} />
      </MockedProvider>
    );
    await new Promise((resolve) => setTimeout(resolve, 0));
    screen.debug();
    expect(
      screen.getByText("Error... Please Refresh The Page")
    ).toBeInTheDocument();
  });
});
