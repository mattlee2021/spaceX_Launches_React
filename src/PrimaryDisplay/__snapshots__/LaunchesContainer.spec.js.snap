// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchesContainer Should display the mocked data 1`] = `
<div
  className="launch__container"
>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Starlink-15 (v1.0)
    </p>
    <p>
      Launch Date: 
      10/24/2020, 11:31:00 AM America/New_York
    </p>
    <p>
      Launch Site: 
      Cape Canaveral Air Force Station Space Launch Complex 40
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Sentinel-6 Michael Freilich
    </p>
    <p>
      Launch Date: 
      11/21/2020, 12:17:00 PM America/New_York
    </p>
    <p>
      Launch Site: 
      Vandenberg Air Force Base Space Launch Complex 4E
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Crew-1
    </p>
    <p>
      Launch Date: 
      11/15/2020, 7:27:00 PM America/New_York
    </p>
    <p>
      Launch Site: 
      Kennedy Space Center Historic Launch Complex 39A
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      GPS III SV04 (Sacagawea)
    </p>
    <p>
      Launch Date: 
      11/5/2020, 6:24:00 PM America/New_York
    </p>
    <p>
      Launch Site: 
      Cape Canaveral Air Force Station Space Launch Complex 40
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Starlink-14 (v1.0)
    </p>
    <p>
      Launch Date: 
      10/24/2020, 11:31:00 AM America/New_York
    </p>
    <p>
      Launch Site: 
      Cape Canaveral Air Force Station Space Launch Complex 40
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Starlink-13 (v1.0)
    </p>
    <p>
      Launch Date: 
      10/18/2020, 8:25:00 AM America/New_York
    </p>
    <p>
      Launch Site: 
      Kennedy Space Center Historic Launch Complex 39A
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Starlink-12 (v1.0)
    </p>
    <p>
      Launch Date: 
      10/6/2020, 7:29:00 AM America/New_York
    </p>
    <p>
      Launch Site: 
      Kennedy Space Center Historic Launch Complex 39A
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Starlink-11 (v1.0)
    </p>
    <p>
      Launch Date: 
      9/3/2020, 8:46:00 AM America/New_York
    </p>
    <p>
      Launch Site: 
      Kennedy Space Center Historic Launch Complex 39A
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      SAOCOM 1B, GNOMES-1, Tyvak-0172
    </p>
    <p>
      Launch Date: 
      8/30/2020, 7:18:00 PM America/New_York
    </p>
    <p>
      Launch Site: 
      Cape Canaveral Air Force Station Space Launch Complex 40
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Starlink-10 (v1.0) & SkySat 19-21
    </p>
    <p>
      Launch Date: 
      8/18/2020, 10:31:00 AM America/New_York
    </p>
    <p>
      Launch Site: 
      Cape Canaveral Air Force Station Space Launch Complex 40
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
  <div
    className="launch__items"
  >
    <p>
      Mission Name: 
      Starlink-9 (v1.0) & BlackSky Global 5-6
    </p>
    <p>
      Launch Date: 
      8/7/2020, 1:12:00 AM America/New_York
    </p>
    <p>
      Launch Site: 
      Kennedy Space Center Historic Launch Complex 39A
    </p>
    <p>
      Rocket Name: 
      Falcon 9
    </p>
    <div
      className="launch__video"
      style={
        Object {
          "height": "360px",
          "width": "640px",
        }
      }
    />
    <form
      className="notes_form"
    >
      <div>
        <label>
          Notes
        </label>
        <br />
        <br />
        <textarea
          className="notes__input"
          onChange={[Function]}
          value=""
        />
      </div>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Copy This Note To Clipboard
      </button>
    </form>
    <hr
      className="line__solid"
    />
  </div>
</div>
`;
